packages/eslint-plugin/src/rules/non-nullable-type-assertion-style.ts

```diff
@@ -43,22 +43,42 @@ export default util.createRule({
       return tsutils.unionTypeParts(type);
     };
 
+    const couldBeNullish = (type: ts.Type): boolean => {
+      if (type.flags & ts.TypeFlags.TypeParameter) {
+        const constraint = type.getConstraint();
+        return constraint == null || couldBeNullish(constraint);
+      } else if (tsutils.isUnionType(type)) {
```

Comment:
I think TypeScript flattens union types internally now, `tsutils.unionTypeParts` is just `return type_1.isUnionType(type) ? type.types : [type];`.

However I'm happy to change it to `tsutils.unionTypeParts` if you think that would be better style.

Reply:
Also note that I can't use `isTypeFlagSet` for union types here, because we also need to check if components of the union are themselves type parameters. Hence the recursive call to `couldBeNullish`.

Reply:
(I could use `isTypeFlagSet` in the `else` branch of the `if` statement though... again happy to do this if you think it's better style.)

Reply:
Aha, I think you understand this are a bit better than me then 😄. I had a false memory of `unionTypeParts` doing more than that. Super! Since you added the test case that showed the code still works as expected with a level of nesting, I'm totally satisfied. 🎉